crates/metadata-providers/src/dynamodb.rs

```diff
@@ -0,0 +1,492 @@
+
+    /// Gets the current version for the given key. If key-value pair is not present, then return
+    /// [`None`].
+    async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError> {
+        let key = self.prefixed(key);
+        let result = self
+            .client
+            .get_item()
+            .table_name(&self.table)
+            .return_consumed_capacity(ReturnConsumedCapacity::None)
+            .attributes_to_get(VERSION_FIELD)
+            .consistent_read(true)
+            .key(KIND_FIELD, AttributeValue::S(key))
+            .send()
+            .await
+            .map_err(ReadError::terminal)?;
+
+        let Some(mut item) = result.item else {
+            return Ok(None);
+        };
+
+        let AttributeValue::N(version_str) = item
+            .remove(VERSION_FIELD)
+            .ok_or(DynamoDbStoreCodecError::MissingField(VERSION_FIELD))?
+        else {
+            return Err(DynamoDbStoreCodecError::InvalidType(VERSION_FIELD).into());
+        };
+
+        let version: u32 = version_str
+            .parse()
+            .map_err(|_| DynamoDbStoreCodecError::InvalidFormat(VERSION_FIELD))?;
+
+        Ok(Some(version.into()))
+    }
+
+    /// Puts the versioned value under the given key following the provided precondition. If the
+    /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
+    async fn put(
+        &self,
+        key: ByteString,
+        value: VersionedValue,
+        precondition: Precondition,
+    ) -> Result<(), WriteError> {
+        let key = self.prefixed(key);
+        let put = self
+            .client
+            .put_item()
+            .table_name(&self.table)
+            .return_consumed_capacity(ReturnConsumedCapacity::None)
+            .return_values(ReturnValue::None)
+            .item(KIND_FIELD, AttributeValue::S(key))
+            .item(
+                VERSION_FIELD,
+                AttributeValue::N(u32::from(value.version).to_string()),
+            )
+            .item(DATA_FIELD, AttributeValue::B(Blob::new(value.value)));
+
+        let put = match precondition {
+            Precondition::None => put,
+            Precondition::DoesNotExist => put.condition_expression(CONDITION_DOES_NOT_EXIST),
+            Precondition::MatchesVersion(version) => put
+                .condition_expression("version = :v")
+                .expression_attribute_values(
+                    ":v",
+                    AttributeValue::N(u32::from(version).to_string()),
+                ),
+        };
+
+        put.send().await.map(|_| ()).map_err(|err| match err {
+            SdkError::ServiceError(err) => match err.into_err() {
+                PutItemError::ConditionalCheckFailedException(exception) => {
+                    WriteError::FailedPrecondition(exception.to_string())
+                }
+                err => WriteError::terminal(err),
+            },
+            err => WriteError::terminal(err),
```

Comment:
Are there errors that are retryable? E.g. `SdkError::TimeoutError`, `SdkError::DispatchError`, `SdkError::ResponseError`?